File: Advanced_kb/src/generics/java/CopyDemo.java
package generics.java;

import generics.java.data.Animal;
import generics.java.data.Cat;

import java.util.ArrayList;
import java.util.List;

class CopyDemo {
    public <T> void copy(List<? super T> dest, List<? extends T> src) {
        dest.add(src.get(0));
    }

    public static void main(String[] args) {
        CopyDemo obj = new CopyDemo();
        List<Cat> cats = new ArrayList<>();
        List<Animal> animals = new ArrayList<>();
        obj.copy(animals, cats);
    }
}

File: Advanced_kb/src/generics/CopyDemo.kt
package generics

import generics.java.data.Animal
import generics.java.data.Cat

fun <T> copyList(dest: MutableList<T>, src: List<T>) {
    src.forEachIndexed { index, element ->
        dest.add(element)
    }
}

fun main() {
    val cats1 = ArrayList<Cat>()
    val cats2 = ArrayList<Cat>()
    val animals1 = ArrayList<Animal>()
    val animals2 = ArrayList<Animal>()

    copyList(cats1, cats2)
    copyList(animals1, cats2)
}